### PR TITLE
PYTHON-3807 add types to mongo_client.py

### PR DESCRIPTION
**Pull request by sleepyStick**
https://jira.mongodb.org/browse/PYTHON-3807

### CODE REVIEW

File: pymongo/collection.py [hunk 1]
```diff
@@ -2250,7 +2248,7 @@ def list_indexes(
         self,
         session: Optional[ClientSession] = None,
         comment: Optional[Any] = None,
-    ) -> CommandCursor[MutableMapping[str, Any]]:
+    ) -> CommandCursor[_DocumentType]:
```
Comment (ShaneHarvey) [inline pymongo/collection.py:2251]: I believe list_indexes always returns MutableMapping. Could you confirm?

_DocumentType is intended to represent documents in the collection, not index documents, so let's avoid using it here.

File: pymongo/collection.py [hunk 2]
```diff
@@ -886,8 +881,11 @@ def _update(
                 comment=comment,
             )
 
-        return self.__database.client._retryable_write(
-            (write_concern or self.write_concern).acknowledged and not multi, _update, session
+        return cast(
+            Mapping[str, Any],
```
Comment (ShaneHarvey) [inline pymongo/collection.py:885]: This cast shouldn't be needed. Is the return type of `_update` inaccurate?

File: pymongo/collection.py [hunk 3]
```diff
@@ -2351,7 +2349,10 @@ def index_information(
         .. versionchanged:: 3.6
            Added ``session`` parameter.
         """
-        cursor = self.list_indexes(session=session, comment=comment)
+        cursor = cast(
+            CommandCursor[MutableMapping[str, Any]],
```
Comment (ShaneHarvey) [inline pymongo/collection.py:2353]: We don't need this cast after we fix list_indexes.

**Review by ShaneHarvey — CHANGES_REQUESTED** [review]

File: bson/codec_options.py [hunk 4]
```diff
@@ -245,6 +245,31 @@ class CodecOptions(Tuple, Generic[_DocumentType]):
...
+        #     document_class: Type[_DocumentType],
+        #     tz_aware: bool = ...,
+        #     uuid_representation: Optional[int] = ...,
+        #     unicode_decode_error_handler: Optional[str] = ...,
+        #     tzinfo: Optional[datetime.tzinfo] = ...,
+        #     type_registry: Optional[TypeRegistry] = ...,
+        #     datetime_conversion: Optional[int] = ...,
+        # ) -> "CodecOptions[_DocumentType]":
+        #     ...
+
```
Comment (ShaneHarvey) [inline bson/codec_options.py:272]: Should this be removed?

**Review by ShaneHarvey — APPROVED** [review]

File: bson/codec_options.py (hunk 4 above)
Comment (sleepyStick) [reply bson/codec_options.py]: yes! thanks for catching that!

**Review by ShaneHarvey — APPROVED** [review]

**Pull request merged**